### PR TITLE
HHH-16130 Make date_trunc function accessible via CriteriaApi

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-16130

### CODE REVIEW

**Comment by gavinking** [comment]
Wait wait wait, I'm -1 on this one.

- `date_trunc` is _not_ on the list of portable HQL functions
- the name `date_trunc` doesn't follow our naming conventions, and
- AFAIK, it's only supported on four databases.

So I don't think we should add it to the criteria API.

**Comment by mbladel** [comment]
I thought it would be nice giving the users the possibility of accessing it through Criteria, since there is no other way of accessing this function (not even the `function()` method can be used, which was a valid workaround in Hibernate 5).
If / when JPA standardizes the concept of a `TemporalUnit` we could consider removing this in favor of that.

If the policy is not to include non-portable HQL functions in Criteria I understand and we can close this.

**Comment by gavinking** [comment]
> If the policy is not to include non-portable HQL functions in Criteria I understand and we can close this.

I think that's pretty much _got_ to be the policy, because without it, where does this end? There's all sorts of nice but unportable SQL functions we could add to `Criteria`.

**Comment by sebersole** [comment]
> not even the function() method can be used, which was a valid workaround in Hibernate 5

Users can register that.  Or even the Dialects

**Comment by beikov** [comment]
How about we emulate truncation (and hence make it portable) by doing something like pseudo-HQL:
```sql
a.ts - mod(a.ts by UNIT, UNIT_RANGE)
```

**Comment by mbladel** [comment]
Updated this since https://github.com/hibernate/hibernate-orm/pull/6134 was merged.

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 1]
```diff
@@ -997,6 +998,9 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
 	@Incubating
 	JpaFunction<Float> second(Expression<? extends TemporalAccessor> datetime);
 
+	@Incubating
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1001]: Java doc would be nice

**Review by beikov — APPROVED** [review]

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1013]: Thanks, added it

**Comment by beikov** [comment]
Any further reservations about adding this, now that we have a portable function implementation @gavinking?

**Comment by gavinking** [comment]
Well currently it's still called `dateTrunc()` instead of  wing an overload of `truncate()`.

**Pull request merged**